keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,945 @@
+4. Non-goal: Create a new empty group and in the same time create new PVCs and add to the new group.
+
+Modify an existing VolumeGroup:
+Add new volume or remove existing volume from an existing VolumeGroup. Option 2 for create VolumeGroup above falls into this case.
+
+Two types of VolumeGroups can be supported:
+
+Type 1: Mutable VolumeGroup - Create new PVC and add to the group.
+* Create a new empty VolumeGroup.
+* Create a new PVC with existing VolumeGroup name. As a result, new PVC is created and added to VolumeGroup. VolumeGroup is modified so Status has this new PVC in PVCList.
+* External-provisioner will be modified so that VolumeGroupName will be passed to the CSI driver when creating a volume.
+
+Only CSI drivers supporting VOLUMEGROUP capability can support this type of mutable volume group.
+When a new PVC is created with the existing VolumeGroup name, the VolumeGroup will be modified and the PVC will be added to PVCList in the Status.
+
+Type 2: Mutable VolumeGroup - Add existing PVC to the group or remove PVC from the group without deleting it. A VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capability will be added to CSI Spec. Only CSI drivers supporting both VOLUMEGROUP and VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capabilities can support this type of mutable volume group.
+* Type 2 is a subset of Type 1 VolumeGroup.
+* Create a new empty VolumeGroup. VolumeGroup Status has a boolean AddRemoveExistingPVC set to true.
+* Add an existing PVC to an existing VolumeGroup (VolumeGroup can be empty to start with or it can have other PVCs already) by adding VolumeGroup name to the PVC Spec.
+* VolumeGroup is modified so the existing PVC is added to the PVCList in the Status.
+* External-provisioner will be modified so that modifying PVC by adding VolumeGroupName will trigger a ModifyVolume call (a new CSI controller RPC) to CSI driver.
+* VolumeGroup controller will be triggered to update the VolumeGroup Status.
+* Deleting a PVC from a VolumeGroup will trigger external-provisioner and the VolumeGroup controller as well.
+
+CSI drivers that do not have a volume_group_id concept can use the VolumeGroup name stored in Kubernetes API server as the volume_group_id.
+
+Snapshot of a VolumeGroup:
+A GroupSnapshot can be created with a VolumeGroup as the source if the CSI driver supports the GROUPSNAPSHOT capability.
+* Create a GroupSnapshot with a VolumeGroup as the source.
+* This will trigger the GroupSnapshot controller to call the CreateGroupSnapshot CSI function and also create multiple VolumeSnapshot API objects with GroupSnapshot name parameter in each VolumeSnapshot Spec. This will trigger the creation of VolumeSnapshotContent API objects and a call to the CreateSnapshot CSI function.
```

Comment:
How would DeleteGroupSnapshot behave? Would it send a single DeleteGroupSnapshot and also individual DeleteSnapshot requests for volumes in the group, letting the CSI driver determine its course of action as detailed in the CreateGroupSnapshot request?

Reply:
Yes, it will send DeleteGroupSnapshot as well as individual DeleteSnapshot requests.
Just like create snapshot, there are 2 cases.
Case 1: Since CSI driver handles individual snapshot creation in CreateGroupSnapshot, it should handle individual snapshot deletion in DeleteGroupSnapshot.
Case 2: Since CSI driver handles individual snapshot creation in CreateSnapshot, it should handle individual snapshot deletion in DeleteSnapshot.